### ISSUE TITLE
Make sbt more friendly to immutable systems

### ISSUE THREAD

**Issue by copumpkin** [sbt/sbt#2498]
For a variety of reasons, mostly related to [Nix](http://nixos.org/nix), I am trying to package sbt in a way where at installation time, it gets a writeable boot directory (the one that usually lives in `~/.sbt/boot`) but during subsequent runs that boot directory is immutable. This is because the Nix store is conceptually append-only (other than a centralized GC) and packages are immutable once installed. To achieve determinism, we don't permit packages to reach into unmanaged locations like `~`.
## steps
1. Create a directory to store the `boot` stuff
2. Give `sbt` an `sbtopts` file that passes `-sbt-dir` and `-sbt-boot` the directory you created in step 1
3. Run `sbt` once to populate the boot directory with a particular sbt revision
4. Make the directory from step 1 read-only
5. Run `sbt` again from a project directory
## problem

`sbt` fails for at least three reasons:
1. It tries to create a `sbt.boot.lock` file, which is unnecessary in this immutable model. I can suppress that by passing the undocumented `boot.lock = true` property into `sbt` in its `sbt.boot.properties` file.
2. It tries to append log entries to an `update.log` file. I can fool sbt into not caring by making `update.log` a symlink to `/dev/null`, but this feels like a hack. Unlike problem 1, I don't think I can tell `sbt` not to use this log file, or configure it to go elsewhere.
3. It tries to create a short-lived `resolution-cache` folder under the boot folder. I haven't figured out how to get past here yet, so there might be other issues lurking
## expectation

`sbt` should be friendlier to this sort of use case. I don't expect direct support, but it would be great if there were an option to turn off the rampant mutation of this folder. Or if not, if there a way to tell the launcher to run from somewhere more static.

**Comment by eed3si9n** [comment]
I like the idea. Having immutable boot probably might have other positive effects like the other issue that you commented on. From the problem description I'm guessing that you would hack a few places in [sbt/launcher](https://github.com/sbt/launcher) to prevent Ivy from running.

**Comment by copumpkin** [comment]
Yeah, basically, although I still need it to pick up the right set of jars. Best would be if I could skip the sbt boot directory altogether and run straight out of the ivy cache, like sbt itself does with projects it manages. Would that be feasible?

**Comment by nrdxp** [comment]
I've just run into this. I've extremely close to declaratively packaging Scala applications in Nix by leveraging `sbt makePom` and [mvn2nix](https://github.com/fzakaria/mvn2nix) to create a local maven repository of the dependencies. Everything is ready to go, but on launch sbt seems to try to fetch itself and scala from the network if the boot directory doesn't exist, which means it fails :cry:

**Comment by jsoo1** [comment]
I started working on an sbt launcher that uses only nix-provided sbt and scala jars. It is buggy and WIP but if anyone here has ideas or can help fix stuff, it is here https://git.sr.ht/~jsoo/single-sbt
